calculator_test.go

```diff
@@ -3,22 +3,196 @@ package calculator_test
 import (
 	"calculator"
 	"testing"
+	"math/rand"
+	"time"
+	
 )
 
 func TestAdd(t *testing.T) {
 	t.Parallel()
-	var want float64 = 4
-	got := calculator.Add(2, 2)
-	if want != got {
-		t.Errorf("want %f, got %f", want, got)
+	type testCase struct {
+		a, b float64
+		want float64
+		message string
 	}
+
+	testCases := []testCase{
+		{a: 2, b: 2, want: 4, message: "Two +ve numbers"},
+		{a: -3, b: 3, want: 0, message: "One +ve and one -ve number"},
+		{a: 5, b: 2, want: 7, message: "Two +ve numbers"},
+	}
+
+	for _, tc := range testCases {
+
+		got := calculator.Add(tc.a, tc.b)
+		if tc.want != got {
+			t.Errorf("%s : Add(%f , %f): want %f, got %f",tc.message, tc.a, tc.b, tc.want, got)
+		}
+	}
+
 }
 
-// func TestSubtract(t *testing.T) {
-// 	t.Parallel()
-// 	var want float64 = 2
-// 	got := calculator.Subtract(4, 2)
-// 	if want != got {
-// 		t.Errorf("want %f, got %f", want, got)
-// 	}
-// }
+func TestAddRandom(t *testing.T) {
+	t.Parallel()
+	rand.Seed(time.Now().UnixNano())
+	for i := 0; i < 10; i++ {
+		a := rand.Float64()
+		b := rand.Float64()
+		want := a + b
+		got := calculator.Add(a, b)
+		if want != got {
+			t.Fatalf("Add(%f, %f): want %f, got %f", a, b, want, got)
+		}
+	}
+}
+
+func TestSubtract(t *testing.T) {
+	t.Parallel()
+	type testCase struct {
+		a, b float64
+		want float64
+		message string
+	}
+
+	testCases := []testCase {
+		{a: 6, b: 5.2, want: 0.7999999999999998, message: "A fractional number in the mix"}, // 0.8 failed the test here. 
```

Comment:
This is a useful thing to know about floating-point numbers: they're inherently imprecise! Comparing `float64` values with `==` is liable to trip you up, so a good alternative is to look at the _difference_ between them. If it's less than some small arbitrary value (say, 0.0001), you can consider them equal. Try writing a function to do this (`CloseEnough`).